File: doomsday/doomsday1.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <algorithm>
#include <climits>
#include <map>
#include <utility>
using namespace std;
#define MAXN 1000
#include <tuple>
char grid[MAXN+2][MAXN+2];
int depth[MAXN + 2][MAXN + 2];

typedef tuple<int,int> tl;
vector< tl > v;

vector < tl > Neighbors(tl k){
	vector < tl > neig;
	int a=get<0>(k); 
	int b=get<1>(k);
	if (grid[a-1][b] != 'x') neig.push_back(tl(a-1,b));
	if (grid[a][b-1] != 'x') neig.push_back(tl(a,b-1));
	if (grid[a+1][b] != 'x') neig.push_back(tl(a+1,b));
	if (grid[a][b+1] != 'x') neig.push_back(tl(a,b+1));
	return neig;
}

int M; 
int N;
void printGrid() {
  for (int i = 1; i <= N; i++) {
    for (int j = 1; j <= M; j++) {
      cout << grid[i][j];
    }
    cout << endl;
  }
}


bool safe() {
  for (int i = 1; i <= N; i++) {
    for (int j = 1; j <= M; j++) {
      if(grid[i][j]=='*'){return false;}
    }
    
  }
return true;
}

int solution(){

	int time=0;
	int flag=1; int flag1=1;
	vector < tl > ne;
	while(flag==1 && flag1==1){	
		printGrid();
		
		for (int i=1; i<=N; i++){
			for (int j=1; j<=M; j++){
				tl k=tl(i,j);
				flag1=0;
				if((grid[i][j]=='+' || grid[i][j]=='-') && depth[i][j]==time){
					flag1=1;				
					vector<tl> neig=Neighbors(k);
					for (tl kl:neig){
						int a=get<0>(kl);
						int b=get<1>(kl);
						if(grid[a][b]!='x' && grid[a][b]=='.'){
								grid[a][b]=grid[i][j];
								depth[a][b]=time+1;
		
						}
						else if (grid[a][b]==grid[i][j]) continue;
						else if ((grid[i][j]=='+' || grid[i][j]=='-') && (grid[a][b]=='+' || grid[a][b]=='-') && grid[a][b]!=grid[i][j]){
						  ne.push_back(tl(a,b));						
							flag=0;
						}
					}
		
				}
			
			}
		}
		time=time+1;
	}
	if (flag==0){
		for (int i=0; i<ne.size(); i++){
			int a=get<0>(ne[i]);
			int b=get<1>(ne[i]);
			grid[a][b]='*';
		}
		cout<<"kaboom"<<'\n';
	}
	return time;
}



int main(){
cin >> N >> M;
for (int i = 1; i <= N; i++){
	for (int j = 1; j <= M; j++) {
      		cin >> grid[i][j];
		depth[i][j]=0;
	}
}



for (int i = 0; i < N; i++) { grid[i][0] = 'x'; grid[i][N + 1] = 'x'; }
for (int i = 0; i < M; i++) { grid[0][i] = 'x'; grid[M + 1][i] = 'x'; }

int time=solution();
cout<<time<<'\n';
if(safe()){
	cout<<"the world is safe"<<'\n';
}
printGrid();
}

File: doomsday/doomsday1.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;
import java.util.Vector;
import java.util.Queue;
import java.io.File;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;
import java.io.FileNotFoundException;
import java.util.NoSuchElementException;
import java.io.BufferedReader;
import java.io.FileReader;

public class pr1 {
	static int MAXN=1000;
	static String[][] grid=new String[MAXN+2][MAXN+2];
	static int[][] depth=new int[MAXN+2][MAXN+2];
	static int M=0,N=0;
	

private static class Pair{
	int x=0;
	int y=0;
void setx(int a){x=a;}
void sety(int a){y=a;}
int getx(){return x;}
int gety(){return y;}	
}

static Queue<Pair> q = new LinkedList<Pair>();
static Queue<Pair> stars = new LinkedList<Pair>();

	public static int min(int a, int b) {
		return a < b ? a : b;
	}

	public static void printGrid(){
		for (int i = 0; i < N; i++) {
    			for (int j = 0; j < M; j++) {
      				System.out.print(grid[i][j]);
    			}
    		System.out.println();
  		}
	}

	public static Vector<Pair> getNeig(int u, int v){
		Vector<Pair> result=new Vector<Pair>();
		if (u > 0 && !grid[ u-1 ][v] .equals( String.valueOf('X'))) {
			Pair pair= new Pair();
			pair.setx(u-1);
			pair.sety(v);
			result.addElement(pair);
		}
  		if (v > 0 && !grid[ u ][v-1] .equals( String.valueOf('X'))){
			Pair pair= new Pair();
			pair.setx(u);
			pair.sety(v-1);
			result.addElement(pair);
		}
  		if (u < N-1 && !grid[u+1] [v] .equals( String.valueOf('X'))){
			Pair pair= new Pair();
			pair.setx(u+1);
			pair.sety(v);
			result.addElement(pair);
		}
  		if (v < M-1 && !grid[u] [v+1] .equals( String.valueOf('X'))){
			Pair pair= new Pair();
			pair.setx(u);
			pair.sety(v+1);
			result.addElement(pair);
		}
  		return result;
	
	}


	public static boolean kaboom(int u, int v, int i, int j){
		 return ((
    		(grid[u][v] .equals( String.valueOf('+')) && grid[i][j].equals( String.valueOf('-'))) ||
    		(grid[u][v].equals( String.valueOf('-')) && grid[i][j] .equals( String.valueOf('+'))))

  		);	
	}

	public static int bfs(){
		int u=0, v=0, i=0, j=0;
		int opt = 34252676;
		while (!q.isEmpty()) {
			Pair pair=q.poll();
			u=pair.getx();
			v=pair.gety();
			if(depth[u][v]==opt){continue;}	
			Vector<Pair> neig=new Vector<Pair>()  ;
			neig=getNeig(u, v);
			for (Pair qq : neig){
				i=qq.getx();
				j=qq.gety();
				if (grid[u][v] .equals( grid[i][j])) {continue;}
			        else if (grid[i][j] .equals( String.valueOf('.'))) {
					depth[i][j] = depth[u][v] + 1;
					grid[i][j] = grid[u][v];
					q.offer(qq);
				}
				else if(kaboom(u,v,i,j)){
					opt=min(opt,depth[u][v]+1);
					stars.offer(qq);				
				}
			}
		}

		while(!stars.isEmpty()){
			Pair pair=stars.poll();
			i=pair.getx();
			j=pair.gety();
			grid[i][j]=String.valueOf('*');	
		}
		return opt;
		
	}
 
public static int textCount(Scanner Scan1) 
        {
            Scanner input = Scan1;
    int lines =0;
    int characters =0;
    int maxCharacters =0;
    String longestLine= "";
     
    while(input.hasNextLine()){
        String line = input.nextLine();
        lines++;
        characters+=line.length();
        if(maxCharacters<line.length()){
            maxCharacters = line.length();
        longestLine = line;
        }
         
    }
	return((characters / lines));
}

	public static void main(String[] args) throws Exception {
		File infile = new File(args[0]);
		Scanner sc = new Scanner(infile);
		sc.useDelimiter("|\\n");
            	M=0;
		N=0;
		BufferedReader reader = new BufferedReader(new FileReader(args[0]));
		int lines = 0;
		while (reader.readLine() != null) lines++;
		reader.close();
		N=lines;
		M=textCount(sc);
		int jj=0;

		sc = new Scanner(infile);
		sc.useDelimiter("|\\n");
		while(jj<N){
			for(int i=0; i<=M; i++){
				String c=sc.next();
				while(c==String.valueOf('\n')){c=sc.next();}

				grid[jj][i]=c;	
		
			}
			jj=jj+1;
		}
printGrid();
	
			for (int i=0; i<N; i++){
				for(int j=0; j<M; j++){				
		
					if(grid[i][j].equals(String.valueOf('+')) || grid[i][j].equals(String.valueOf('-'))){
						Pair pair= new Pair();
						pair.setx(i);
						pair.sety(j);
						q.add(pair);
						depth[i][j]=0;
					}
					else {depth[i][j]=-1;}
				}
			}
			
		



int opt=bfs();
if (opt==34252676){
	System.out.println("the world is saved");
}
else{
	System.out.println (opt);}
printGrid();




  
   }
}
